src/Array.mo

```diff
@@ -7,123 +17,327 @@ import Prim "mo:⛔";
+  /// list of elements). Uses `equal` to compare elements in the arrays.
+  ///
+  /// ```motoko include=import
+  /// // Use the equal function from the Nat module to compare Nats
+  /// import {equal} "mo:base/Nat"; 
+  ///
+  /// let array1 = [0, 1, 2, 3];
+  /// let array2 = [0, 1, 2, 3];
+  /// Array.equal(array1, array2, equal)
+  /// ```
+  ///
+  /// Runtime: O(size1 + size2)
+  ///
+  /// Space: O(1)
+  /// 
+  /// *Runtime and space assumes that `equal` runs in O(1) time and space.
+  public func equal<X>(array1 : [X], array2 : [X], equal : (X, X) -> Bool) : Bool {
+    let size1 = array1.size();
+    let size2 = array2.size();
+    if (size1 != size2) {
       return false;
     };
     var i = 0;
-    while (i < a.size()) {
-      if (not eq(a[i], b[i])) {
+    while (i < size1) {
+      if (not equal(array1[i], array2[i])) {
         return false;
       };
       i += 1;
     };
     return true;
   };
 
-  /// Append the values of two input arrays
+  /// Returns first value in `array` for which `predicate` returns true.
+  /// If no element satisfies the predicate, returns null.
+  ///
+  /// ```motoko include=import
+  /// let array = [1, 9, 4, 8];
+  /// Array.find<Nat>(array, func x = x > 8)
+  /// ```
+  /// Runtime: O(size)
+  ///
+  /// Space: O(1)
+  /// 
+  /// *Runtime and space assumes that `predicate` runs in O(1) time and space.
+  public func find<X>(array : [X], predicate : X -> Bool) : ?X {
+    for (element in array.vals()) {
+      if (predicate element) {
+        return ?element;
+      }
+    };
+    return null;
+  };
+
+  /// Create a new array by appending the values of `array1` and `array2`.
   /// @deprecated `Array.append` copies its arguments and has linear complexity; when used in a loop, consider using a `Buffer`, and `Buffer.append`, instead.
-  public func append<A>(xs : [A], ys : [A]) : [A] {
-    switch(xs.size(), ys.size()) {
-      case (0, 0) { []; };
-      case (0, _) { ys; };
-      case (_, 0) { xs; };
-      case (xsSize, ysSize) {
-        Prim.Array_tabulate<A>(xsSize + ysSize, func (i : Nat) : A {
-          if (i < xsSize) {
-            xs[i];
-          } else {
-            ys[i - xsSize];
-          };
-        });
+  ///
+  /// ```motoko include=import
+  /// let array1 = [1, 2, 3];
+  /// let array2 = [4, 5, 6];
+  /// Array.append<Nat>(array1, array2)
+  /// ```
+  /// Runtime: O(size1 + size2)
+  ///
+  /// Space: O(size1 + size2)
+  public func append<X>(array1 : [X], array2 : [X]) : [X] {
+    let size1 = array1.size();
+    let size2 = array2.size();
+    Prim.Array_tabulate<X>(size1 + size2, func i {
+      if (i < size1) {
+        array1[i];
+      } else {
+        array2[i - size1];
       };
-    };
+    });
   };
 
-  /// Sorts the given array, in ascending order, according to the `compare` function.
-  /// This is a _stable_ sort.
+  // FIXME this example stack overflows. Should test with new implementation of sortInPlace
```

Comment:
Why does this example cause a stack overflow?

Reply:
Only in the documentation, which uses the interpreter.

https://github.com/dfinity/motoko/pull/3541#discussion_r1014351972